### PR TITLE
ADBDEV-2885: Postgres planner produces bogus plan for query to replicated table with SIRV function

### PR DESCRIPTION
**Pull request by RekGRpth**
Query with subplan containing volatile functions on distributed replicated tables does not make motion gather. This produces wrong plan and leads to segfault.
Steps to reproduce this problem:
```sql
create table t (i int) distributed replicated;
insert into t select 1;
create function f(i int) returns int language sql security definer as $$ select i; $$;
select (select f(i) from t); -- <-- segfault here!!!
select (select f(i) from t group by f(i)); -- <-- segfault here!!!
select (select i from t group by i having f(i) > 0); -- <-- segfault here!!!
```
Wrong plans:
```sql
explain (costs off, verbose) SELECT (select f(i) from t);
             QUERY PLAN              
-------------------------------------
 Result
   Output: $0
   InitPlan 1 (returns $0)
     ->  Seq Scan on public.t
           Output: f(i)
 Optimizer: Postgres query optimizer
(6 rows)
explain (costs off, verbose) SELECT (select f(i) from t group by f(i));
             QUERY PLAN              
-------------------------------------
 Result
   Output: $0
   InitPlan 1 (returns $0)
     ->  HashAggregate
           Output: (f(i))
           Group Key: f(t.i)
           ->  Seq Scan on public.t
                 Output: i, f(i)
 Optimizer: Postgres query optimizer
(9 rows)
explain (costs off, verbose) SELECT (select i from t group by i having f(i) > 0);
             QUERY PLAN              
-------------------------------------
 Result
   Output: $0
   InitPlan 1 (returns $0)
     ->  HashAggregate
           Output: i
           Group Key: t.i
           Filter: (f(t.i) > 0)
           ->  Seq Scan on public.t
                 Output: i
 Optimizer: Postgres query optimizer
(10 rows)
```
Solving problem includes make explicit motion gather for such subplan and only for it.
Right plans:
```sql
explain (costs off, verbose) SELECT (select f(i) from t);
                    QUERY PLAN                    
--------------------------------------------------
 Result
   Output: $0
   InitPlan 1 (returns $0)  (slice2)
     ->  Gather Motion 1:1  (slice1; segments: 1)
           Output: (f(i))
           ->  Seq Scan on public.t
                 Output: f(i)
 Optimizer: Postgres query optimizer
(8 rows)
explain (costs off, verbose) SELECT (select f(i) from t group by f(i));
                    QUERY PLAN                    
--------------------------------------------------
 Result
   Output: $0
   InitPlan 1 (returns $0)  (slice2)
     ->  Gather Motion 1:1  (slice1; segments: 1)
           Output: (f(i))
           ->  HashAggregate
                 Output: (f(i))
                 Group Key: f(t.i)
                 ->  Seq Scan on public.t
                       Output: i, f(i)
 Optimizer: Postgres query optimizer
(11 rows)
explain (costs off, verbose) SELECT (select i from t group by i having f(i) > 0);
                    QUERY PLAN                    
--------------------------------------------------
 Result
   Output: $0
   InitPlan 1 (returns $0)  (slice2)
     ->  Gather Motion 1:1  (slice1; segments: 1)
           Output: i
           ->  HashAggregate
                 Output: i
                 Group Key: t.i
                 Filter: (f(t.i) > 0)
                 ->  Seq Scan on public.t
                       Output: i
 Optimizer: Postgres query optimizer
(12 rows)
```

### CODE REVIEW

File: src/backend/cdb/cdbllize.c [hunk 1]
```diff
@@ -1106,7 +1106,9 @@ focusPlan(Plan *plan, bool stable, bool rescannable)
 	 * do nothing.
 	 */
 	if (plan->flow->flotype == FLOW_SINGLETON &&
-		plan->flow->locustype != CdbLocusType_SegmentGeneral)
+		plan->flow->locustype != CdbLocusType_SegmentGeneral &&
+		(plan->dispatch != DISPATCH_PARALLEL ||
+		!contain_volatile_functions((Node *)plan->targetlist)))
```
Comment (Stolb27) [inline src/backend/cdb/cdbllize.c]: Please describe the meaning of both code lines in the commit and PR description.

File: src/test/regress/expected/rpt.out [hunk 2]
```diff
@@ -1241,6 +1241,40 @@ select c from rep_tab where c in (select distinct d from rand_tab);
  2
 (2 rows)
 
+set optimizer= off;
```
Comment (Stolb27) [inline src/test/regress/expected/rpt.out]: Please provide test description.
It's not recommended to limit test execution by some planner (except cases when it leads to creation new copy of out file, or reproduction steps uses fault injector and freezes on the different optimizer).

File: src/test/regress/expected/rpt.out [hunk 3]
```diff
@@ -1241,6 +1241,40 @@ select c from rep_tab where c in (select distinct d from rand_tab);
...
+drop table if exists t;
+NOTICE:  table "t" does not exist, skipping
+create table t (i int) distributed replicated;
+insert into t select 1;
+create or replace function f(i int) returns int language sql security definer as $$ select i; $$;
+explain SELECT (select f(i) from t);
```
Comment (Stolb27) [inline src/test/regress/expected/rpt.out]: You may use `costs off` to decrease patch footprint in this case. More, I think that explain analyze is enough (valid plan + SIGSEGV absence).

Comment (RekGRpth) [reply src/test/regress/expected/rpt.out]: done

File: src/backend/cdb/cdbllize.c (hunk 1 above)
Comment (RekGRpth) [reply src/backend/cdb/cdbllize.c]: for query (on distributed replicated tables) containing volatile functions consider adjust plan flow

File: src/test/regress/expected/rpt.out (hunk 2 above)
Comment (RekGRpth) [reply src/test/regress/expected/rpt.out]: done

**Comment by Stolb27** [comment]
The next queries are planned incorrect with current version of patch (SIGSEGV without patch):

```
SELECT (select f(i) from t group by f(i));
SELECT (select i from t group by i having f(i) > 0);
```

In my opinion, you are trying to solve exact problem, but the problem is wider. ~~Replicated table implementation in gpdb should be rewritten from scratch~~. GPDB refuses to apply motion for data outside the coordinator.

File: src/include/nodes/primnodes.h [hunk 4]
```diff
@@ -1506,6 +1506,7 @@ typedef struct Flow
     /* The original Flow ptr is saved here upon setting req_move. */
     struct Flow    *flow_before_req_move;
 
+	bool focus_need;
```
Comment (HustonMmmavr) [inline src/include/nodes/primnodes.h:1511]: Maybe it's worth to add commentary about this flag? Also, seems, here is a problem with indentation.

Comment (RekGRpth) [reply src/include/nodes/primnodes.h:1511]: > problem with indentation

I only add one line here and do not touch other lines and indentations.

Comment (RekGRpth) [reply src/include/nodes/primnodes.h:1511]: > Maybe it's worth to add commentary about this flag?

done

**Review by HustonMmmavr — DISMISSED** [review]

**Pull request closed without merging**